### PR TITLE
fix(levm): change blobhash fee calculation 

### PR DESCRIPTION
**Pull request by lima-limon-inc**
**Motivation**
The way we calculate the blob fee is not following the specification; which meant over charging the sender.
Also, this changes the way we return a 0 in case the index is not valid (that is, when the index is larger than the blob hash array itself).

**Description**

We are using the `tx.max_fee_per_blob_gas` when the specification uses the `base_fee_per_blob_gas`.  For this, we use the already available function `get_base_fee_per_blob_gas`. 

Also, now a 0 is pushed to the stack when the index is larger than usize when it reality it should check against the amount of blobs. The previous implementation achieved this, but this new implementation is more explicit about _when_ it does it. 

Closes https://github.com/lambdaclass/ethrex/issues/1508

### CODE REVIEW

**Comment by JereSalo** [comment]
Sorry, I don't see what was wrong with the previous implementation:
```rust
pub fn op_blobhash(
        &mut self,
        current_call_frame: &mut CallFrame,
    ) -> Result<OpcodeSuccess, VMError> {
        self.increase_consumed_gas(current_call_frame, gas_cost::BLOBHASH)?;

        let Ok(index) = TryInto::<usize>::try_into(current_call_frame.stack.pop()?) else {
            current_call_frame.stack.push(U256::zero())?;
            return Ok(OpcodeSuccess::Continue);
        };

        let blob_hashes = &self.env.tx_blob_hashes;

        blob_hashes
            .get(index)
            .map(|el| {
                current_call_frame
                    .stack
                    .push(U256::from_big_endian(el.as_bytes()))
            })
            .unwrap_or_else(|| current_call_frame.stack.push(U256::zero()))?;

        Ok(OpcodeSuccess::Continue)
    }
```
If index is too big for `usize` then 0 is pushed to the stack.
Otherwise we try to get the blob hash in `index` position, but if there is none then we push 0 to the stack.
Isn't this behavior right?

**Comment by lima-limon-inc** [comment]
Good morning Jere!

> (...) we try to get the blob hash in `index` position, but if there is none then we push 0 to the stack. Isn't this behavior right?

I believe this behavior is correct. However, I think the problem lies on the first part:

> If index is too big for `usize` then 0 is pushed to the stack. 

AFAIAA, that behavior is not in the specification. Do you know if that is intended? [Here's](https://eips.ethereum.org/EIPS/eip-4844#opcode-to-get-versioned-hashes) the EIP where this Operand is introduced. 
Here it states:

> and replaces it on the stack with tx.blob_versioned_hashes[index] if index < len(tx.blob_versioned_hashes), and otherwise with a zeroed bytes32 value. 

I _think_ this means that 0 should be pushed only when there are not enough items on the blobhash array. I don't think other types of errors (such as casting errors) also imply pushing 0; although the wording is a bit vague. What do you make of this?


Maybe the correct behavior would be to retung an internal error?


PS: I missed the `.unwrap_or_else(|| current_call_frame.stack.push(U256::zero()))?;` bit. If we leave the checking outside the map function, then maybe we could return an internal error here as well. Seen in how the index being valid is already checked.

What do you think?

File: crates/vm/levm/src/vm.rs [hunk 1]
```diff
@@ -589,14 +589,13 @@ impl VM {
...
-        let blob_gas_cost = self
-            .env
-            .tx_max_fee_per_blob_gas
-            .unwrap_or_default()
-            .checked_mul(blob_gas_used)
+        let base_fee_per_blob_gas = self.get_base_fee_per_blob_gas()?;
+
+        let blob_fee = blob_gas_used
+            .checked_mul(base_fee_per_blob_gas)
             .ok_or(InternalError::UndefinedState(1))?;
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/vm.rs:596]: This seemed to be the issue. The [EIP-4844](https://eips.ethereum.org/EIPS/eip-4844#gas-accounting) describes how to the gas for the blob hashes should be computed.

Before this patch, it used the max fee per blob gas to calculate the value. However, that makes me wonder where the max fee per blob gas is used.

Regardless, more tests seem to pass(?

File: crates/vm/levm/src/opcode_handlers/block.rs [hunk 2]
```diff
@@ -160,12 +160,14 @@ impl VM {
...
-            return Ok(OpcodeSuccess::Continue);
-        };
+        let index = current_call_frame.stack.pop()?;
 
         let blob_hashes = &self.env.tx_blob_hashes;
+        if index > blob_hashes.len().into() {
+            current_call_frame.stack.push(U256::zero())?;
+            return Ok(OpcodeSuccess::Continue);
+        }
+        let index: usize = index.try_into().map_err(|_| VMError::VeryLargeNumber)?;
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/opcode_handlers/block.rs:171]: I also left this change. I think it's more explicit and follows a bit closer the reference implementation

File: crates/vm/levm/src/opcode_handlers/block.rs [hunk 3]
```diff
@@ -160,12 +160,14 @@ impl VM {
...
```
[the remark this answers is not in the document]
Comment (lima-limon-inc) [reply crates/vm/levm/src/opcode_handlers/block.rs:170]: Indeed, I'll change it right away.

[the remark this answers is not in the document]
Comment (lima-limon-inc) [reply crates/vm/levm/src/opcode_handlers/block.rs]: Changed in https://github.com/lambdaclass/ethrex/pull/1509/commits/f0525867181d0fc3074fedb4264e2a1a7a25ae97

File: crates/vm/levm/src/opcode_handlers/block.rs [hunk 4]
```diff
@@ -160,21 +160,27 @@ impl VM {
...
+            .map_err(|_| VMError::Internal(InternalError::ConversionError))?;
 
-        blob_hashes
-            .get(index)
-            .map(|el| {
-                current_call_frame
-                    .stack
-                    .push(U256::from_big_endian(el.as_bytes()))
-            })
-            .unwrap_or_else(|| current_call_frame.stack.push(U256::zero()))?;
```
Comment (lima-limon-inc) [inline crates/vm/levm/src/opcode_handlers/block.rs:177]: @JereSalo Do you know why we were using map here? I didn't quite understand it.

Were we mapping the Some(value) and doing nothing if it was none?

File: crates/vm/levm/src/opcode_handlers/block.rs [hunk 5]
```diff
@@ -160,21 +160,27 @@ impl VM {
...
+        if let Some(blob_hash) = blob_hashes.get(index) {
+            current_call_frame
+                .stack
+                .push(U256::from_big_endian(blob_hash.as_bytes()))?;
+        } else {
+            // This should never happen because whe check if the index
+            // fits inside the blobhash array above
+            return Err(VMError::Internal(InternalError::UndefinedState(1)));
+        }
```
Comment (JereSalo) [inline crates/vm/levm/src/opcode_handlers/block.rs:183]:
```suggestion
        let blob_hash = blob_hashes
    .get(index)
    .ok_or_else(|| VMError::Internal(InternalError::UndefinedState(1)))?;

current_call_frame
    .stack
    .push(U256::from_big_endian(blob_hash.as_bytes()))?;
```
IMO this looks better.
We don't want to use UndefinedState internal error inside opcodes, that is reserved for special occasions. Change it for another `InternalError`, you can add one for when we try to access a list with an index that is out of bounds.

**Review by JereSalo — CHANGES_REQUESTED** [review]

File: crates/vm/levm/src/opcode_handlers/block.rs (hunk 5 above)
Comment (lima-limon-inc) [reply crates/vm/levm/src/opcode_handlers/block.rs:183]: Noted Jere, I'll change it ASAP!

Comment (lima-limon-inc) [reply crates/vm/levm/src/opcode_handlers/block.rs]: Changed here: https://github.com/lambdaclass/ethrex/pull/1509/commits/b0613f9f9200a77ab0b79ffc541d8bf9267a5c31

**Review by JereSalo — APPROVED** [review]

**Pull request merged**